File: packages/flutter/lib/src/internal/wheel.dart
/// The wheel over a strip that runs the other way, and what happens when the
/// strip runs out of room.
library;

import 'package:flutter/gestures.dart';
import 'package:flutter/widgets.dart';
import 'package:plass_ui/src/types.dart';

/// How long after the strip last moved the wheel still belongs to it.
///
/// Long enough to cover the gap between two notches of a wheel being turned,
/// short enough that a reader who has stopped and looked at what arrived is
/// scrolling the page again by the time they reach for it.
const Duration _latch = Duration(milliseconds: 250);

/// A pixel of slack, for the fractional layout that reports a strip everything
/// fits in as a hair longer than its box.
const double _epsilon = 1;

/// The scroll signals a strip takes off the framework, and the ones it refuses.
///
/// Two widgets have the same problem and it is not the one it looks like. A
/// mouse has one wheel, it points down the page, and a strip that runs across
/// the box has no use for that direction: a horizontal [Scrollable] reads the
/// horizontal half of a scroll and a mouse only ever produces the vertical one,
/// so the shelf under the pointer sits still while whatever is behind it moves
/// instead. The pointer being on the strip is the reader saying which of the two
/// things under it they meant to move, and [turn] is what acts on that.
///
/// [overscroll] is the second half, and it is the end of the strip rather than
/// the middle of it. A scroller that has run out hands the rest of the gesture
/// to whatever is behind it, and the reader who was flicking a shelf along gets
/// the whole page moving instead — without having asked for it, and usually
/// without having seen which pixel it happened at.
///
/// Everything goes through the [PointerSignalResolver] rather than being acted
/// on where it arrives. A scroll view that has claimed the same event is deeper
/// in the hit test and registers first, so the strip's own [Scrollable] moves it
/// and this does not move it again; and a claim that does nothing at all is what
/// keeps an ancestor from taking a signal the strip has decided to hold.
class PlassWheelScroll extends StatefulWidget {
  /// Wraps [child] so the strip [controller] drives reads the wheel.
  const PlassWheelScroll({
    required this.controller,
    required this.overscroll,
    required this.child,
    this.turn = true,
    super.key,
  });

  /// The strip's own controller. Nothing happens before it has a position.
  final ScrollController controller;

  /// Whether a wheel that points across the strip moves it along.
  ///
  /// The axis the framework has no answer for. A trackpad's two fingers and a
  /// tilt wheel already produce a gesture along the strip and are left alone.
  final bool turn;

  /// What the strip does with a gesture it has run out of room for.
  final PlassOverscroll overscroll;

  /// The strip.
  final Widget child;

  @override
  State<PlassWheelScroll> createState() => _PlassWheelScrollState();
}

class _PlassWheelScrollState extends State<PlassWheelScroll> {
  /// When the strip last had somewhere to go, which is what the latch reads.
  Duration? _moved;

  void _onPointerSignal(PointerSignalEvent event) {
    if (event is! PointerScrollEvent || !widget.controller.hasClients) {
      return;
    }

    final ScrollPosition position = widget.controller.position;
    final bool horizontal = position.axis == Axis.horizontal;
    final Offset delta = event.scrollDelta;
    final double along = horizontal ? delta.dx : delta.dy;
    final double across = horizontal ? delta.dy : delta.dx;

    // A gesture that already points along the strip is the framework's own, and
    // the [Scrollable] inside this listener has registered for it already.
    final bool own = along.abs() >= across.abs();
    final double travel = own ? along : across;

    if (travel == 0) {
      return;
    }

    // The sentence the whole containment rests on: a strip everything fits in
    // is not a scroller, so it never takes the wheel and never holds it.
    // Without this, a bar of three tabs would be a place on the page the reader
    // cannot scroll past.
    if (position.maxScrollExtent - position.minScrollExtent <= _epsilon) {
      return;
    }

    final double target = (position.pixels + travel).clamp(
      position.minScrollExtent,
      position.maxScrollExtent,
    );

    if (target != position.pixels) {
      _moved = event.timeStamp;

      // Somewhere to go, and along the strip's own axis the framework is
      // already taking it there.
      if (own || !widget.turn) {
        return;
      }

      GestureBinding.instance.pointerSignalResolver.register(event, (PointerSignalEvent _) {
        widget.controller.jumpTo(target);
      });

      return;
    }

    // Nothing left this way. A strip the wheel does not turn has no claim on the
    // cross axis at all, so there is nothing for it to hold there either.
    if (!own && !widget.turn) {
      return;
    }

    final Duration? moved = _moved;
    final bool latched = moved != null && event.timeStamp - moved < _latch;

    // `contain` keeps the gesture; `auto` gives it back, but not in the middle
    // of the flick that was moving the strip a moment ago.
    if (widget.overscroll == PlassOverscroll.contain || latched) {
      // Claimed and dropped, which is the whole of what containment is: the
      // signal is spoken for, so nothing behind the strip acts on it.
      GestureBinding.instance.pointerSignalResolver.register(event, (PointerSignalEvent _) {});
    }
  }

  @override
  Widget build(BuildContext context) {
    if (!widget.turn && widget.overscroll == PlassOverscroll.auto) {
      return widget.child;
    }

    return Listener(onPointerSignal: _onPointerSignal, child: widget.child);
  }
}

File: packages/react/src/internal/wheel.ts
'use client';

/**
 * The wheel over a strip that runs the other way, and what happens when the
 * strip runs out of room.
 *
 * Two components have the same problem and it is not the one it looks like. A
 * mouse has one wheel, it points down the page, and a strip that runs across
 * the box has no use for that direction — so what the wheel does over a shelf
 * of cards or a bar of tabs is left to the browser, which is to say it is one
 * thing on one machine and another on the next. The pointer being on the strip
 * is the reader saying which of the two things under it they meant to move, and
 * this is what acts on that.
 *
 * The second half is what happens at the end of the strip. A nested scroller
 * that runs out hands the rest of the gesture to the page, and the reader who
 * was flicking a shelf along gets the whole page moving instead, without having
 * asked for it and usually without having noticed which pixel it happened at.
 * `overscroll` is the choice between the two answers, and neither of them is
 * "sometimes": see `PlassOverscroll`.
 *
 * Only the axis the browser has no answer for. A trackpad's two fingers, a tilt
 * wheel and Shift held down all produce a horizontal gesture, which scrolls a
 * horizontal strip on its own — and chains on its own, which is what
 * `overscroll-behavior-x` is for. Nothing here touches those.
 */

import * as React from 'react';
import type { PlassOverscroll } from '../types.js';

/**
 * What one line is worth in pixels, for the browsers that report a wheel in
 * lines rather than in pixels.
 */
const WHEEL_LINE = 16;

/**
 * How long after the strip last moved the wheel still belongs to it.
 *
 * The window a browser latches a nested scroller for, near enough: long enough
 * to cover the gap between two notches of a wheel being turned, short enough
 * that a reader who has stopped and looked at what arrived is scrolling the
 * page again by the time they reach for it.
 */
const LATCH_MS = 250;

export interface WheelScrollOptions {
  /** Whether the wheel is read at all. A strip that runs down the page is not. */
  enabled: boolean;
  /** What happens once the strip has nowhere left to go. */
  overscroll: PlassOverscroll;
}

/**
 * Turns a vertical wheel over a horizontal scroller into travel along it.
 *
 * A native listener rather than `onWheel`, because React attaches its own wheel
 * listener to the root passively, and `preventDefault` inside a passive
 * listener does nothing but log.
 */
export function useWheelScroll(
  ref: React.RefObject<HTMLElement | null>,
  { enabled, overscroll }: WheelScrollOptions
): void {
  React.useEffect(() => {
    const element = ref.current;

    if (!element || !enabled) {
      return;
    }

    /**
     * When the strip last moved on a wheel, which is what the latch reads.
     *
     * Never, to begin with. A `0` would be a move at the moment the page
     * started, so for the first quarter-second of a page's life a strip that
     * had never moved would hold the wheel anyway.
     */
    let moved = Number.NEGATIVE_INFINITY;

    const onWheel = (event: WheelEvent) => {
      // A gesture that already has a horizontal half is one the browser scrolls
      // the strip with by itself, and a second handler would double it.
      if (Math.abs(event.deltaY) <= Math.abs(event.deltaX)) {
        return;
      }

      const room = element.scrollWidth - element.clientWidth;

      // A pixel of slack, and the sentence the whole containment rests on: a
      // strip everything fits in is not a scroller, so it never takes the wheel
      // and never holds it. Without this, a bar of three tabs would be a place
      // on the page the reader cannot scroll past.
      if (room <= 1) {
        return;
      }

      const distance =
        event.deltaMode === event.DOM_DELTA_LINE
          ? event.deltaY * WHEEL_LINE
          : event.deltaMode === event.DOM_DELTA_PAGE
            ? event.deltaY * element.clientWidth
            : event.deltaY;

      // `abs`, because a right-to-left container counts its scroll backwards
      // from zero. How far along we are is a distance either way.
      const along = Math.abs(element.scrollLeft);
      const left = distance > 0 ? room - along > 1 : along > 1;

      if (!left) {
        // Nothing left this way. `contain` keeps the wheel anyway; `auto` gives
        // it back, but not in the middle of the gesture that was scrolling the
        // strip a moment ago.
        if (overscroll === 'contain' || event.timeStamp - moved < LATCH_MS) {
          event.preventDefault();
        }

        return;
      }

      // Which way "forward" is on the physical axis: flipped under RTL.
      const sign = getComputedStyle(element).direction === 'rtl' ? -1 : 1;

      moved = event.timeStamp;
      event.preventDefault();
      element.scrollBy({ left: distance * sign, behavior: 'auto' });
    };

    element.addEventListener('wheel', onWheel, { passive: false });

    return () => element.removeEventListener('wheel', onWheel);
  }, [enabled, overscroll, ref]);
}

/**
 * The class that keeps a scroller's *own* axis to itself.
 *
 * The other half of `overscroll`, and it is CSS rather than JavaScript because
 * this axis is one the browser already scrolls: a two-finger swipe along the
 * shelf, and on a Mac the swipe past the end of it that goes back a page.
 *
 * One axis, never both. A horizontal strip that also contained the vertical
 * axis would be a box a finger cannot scroll the page from, which is most of
 * the screen on a phone.
 */
export function overscrollClasses(overscroll: PlassOverscroll, horizontal: boolean): string {
  if (overscroll !== 'contain') {
    return '';
  }

  return horizontal ? 'overscroll-x-contain' : 'overscroll-y-contain';
}
